Description: Display locale @modifiers properly
 Make items in the UI represented by locale modifiers distinguishable from
 items without modifiers.

 The commenting of the language_name_get_codeset_details() function is
 unrelated to the modifiers, and is motivated like this in the changelog:
 "Don't mess with non-utf8 locales, or else the language chooser in
  the User Accounts UI may display confusing info (LP: #864618)."
 https://launchpad.net/ubuntu/+source/gnome-control-center/1:3.2.1-0ubuntu1
 We don't set language any longer when adding users, but OTOH it's still
 true that we don't want non-utf8 locales show up in the UIs. It stays
 for now.
Author: Gunnar Hjalmarsson <gunnarhj@ubuntu.com>
Date: Tue, 26 Mar 2013 17:04:18 +0100
Forwarded: https://gitlab.gnome.org/GNOME/gnome-desktop/issues/50

 {a => b}/libgnome-desktop/gnome-languages.c | 82 +++++++++++++++++++++++++++--
 {a => b}/libgnome-desktop/gnome-languages.h |  2 +
 2 files changed, 81 insertions(+), 3 deletions(-)

diff --git a/libgnome-desktop/gnome-languages.c b/libgnome-desktop/gnome-languages.c
index 53cd1bf..9b3b68b 100644
--- a/libgnome-desktop/gnome-languages.c
+++ b/libgnome-desktop/gnome-languages.c
@@ -1082,6 +1082,8 @@ gnome_get_language_from_locale (const char *locale,
         g_autofree char *langinfo_codeset = NULL;
         g_autofree char *translated_language = NULL;
         g_autofree char *translated_territory = NULL;
+        g_autofree char *modifier = NULL;
+        g_autofree char *translated_modifier = NULL;
         gboolean is_utf8 = TRUE;
 
         g_return_val_if_fail (locale != NULL, NULL);
@@ -1096,7 +1098,7 @@ gnome_get_language_from_locale (const char *locale,
                             &language_code,
                             &territory_code,
                             &codeset_code,
-                            NULL);
+                            &modifier);
 
         if (language_code == NULL) {
                 goto out;
@@ -1122,7 +1124,7 @@ gnome_get_language_from_locale (const char *locale,
                                         translated_territory);
         }
 
-        language_name_get_codeset_details (locale, &langinfo_codeset, &is_utf8);
+        //language_name_get_codeset_details (locale, &langinfo_codeset, &is_utf8);
 
         if (codeset_code == NULL && langinfo_codeset != NULL) {
                 codeset_code = g_strdup (langinfo_codeset);
@@ -1134,6 +1136,13 @@ gnome_get_language_from_locale (const char *locale,
                                         codeset_code);
         }
 
+        if (modifier != NULL) {
+                translated_modifier = gnome_get_translated_modifier (modifier, translation);
+        }
+        if (translated_modifier != NULL) {
+                g_string_append_printf (full_language, " - %s", translated_modifier);
+        }
+
  out:
         if (full_language->len == 0) {
                 g_string_free (full_language, TRUE);
@@ -1167,6 +1176,8 @@ gnome_get_country_from_locale (const char *locale,
         g_autofree char *langinfo_codeset = NULL;
         g_autofree char *translated_language = NULL;
         g_autofree char *translated_territory = NULL;
+        g_autofree char *modifier = NULL;
+        g_autofree char *translated_modifier = NULL;
         gboolean is_utf8 = TRUE;
 
         g_return_val_if_fail (locale != NULL, NULL);
@@ -1181,7 +1192,7 @@ gnome_get_country_from_locale (const char *locale,
                             &language_code,
                             &territory_code,
                             &codeset_code,
-                            NULL);
+                            &modifier);
 
         if (territory_code == NULL) {
                 goto out;
@@ -1194,6 +1205,13 @@ gnome_get_country_from_locale (const char *locale,
 		goto out;
 	}
 
+        if (modifier != NULL) {
+                translated_modifier = gnome_get_translated_modifier (modifier, translation);
+        }
+        if (translated_modifier != NULL) {
+                g_string_append_printf (full_name, " - %s", translated_modifier);
+        }
+
         if (language_code != NULL) {
                 translated_language = get_translated_language (language_code, translation);
         }
@@ -1308,6 +1326,64 @@ gnome_get_country_from_code (const char *code,
         return get_translated_territory (code, translation);
 }
 
+/**
+ * gnome_get_translated_modifier:
+ * @modifier: the modifier part of a locale name
+ * @translation: (allow-none): a locale string
+ *
+ * Gets a translation of the raw @modifier string. If @translation
+ * is provided the returned string is translated accordingly.
+ *
+ * Return value: (transfer full): the translated modifier string.
+ * Caller takes ownership.
+ *
+ * Since: 3.34
+ */
+char *
+gnome_get_translated_modifier (const char *modifier,
+                               const char *translation)
+{
+        g_return_val_if_fail (modifier != NULL, NULL);
+
+        const char *retval;
+        GHashTable *modifiers_map = g_hash_table_new (g_str_hash, g_str_equal);
+
+        g_hash_table_insert (modifiers_map, g_strdup ("abegede"), g_strdup (_("Abegede")));
+        g_hash_table_insert (modifiers_map, g_strdup ("cyrillic"), g_strdup (_("Cyrillic")));
+        g_hash_table_insert (modifiers_map, g_strdup ("devanagari"), g_strdup (_("Devanagari")));
+        g_hash_table_insert (modifiers_map, g_strdup ("iqtelif"), g_strdup (_("IQTElif")));
+        g_hash_table_insert (modifiers_map, g_strdup ("latin"), g_strdup (_("Latin")));
+        g_hash_table_insert (modifiers_map, g_strdup ("saaho"), g_strdup (_("Saho")));
+        g_hash_table_insert (modifiers_map, g_strdup ("valencia"), g_strdup (_("Valencia")));
+
+        if (g_hash_table_contains (modifiers_map, modifier)) {
+                locale_t loc;
+                locale_t old_locale;
+
+                if (translation != NULL) {
+                        loc = newlocale (LC_MESSAGES_MASK, translation, (locale_t) 0);
+                        if (loc == (locale_t) 0) {
+                                g_hash_table_destroy (modifiers_map);
+                                return NULL;
+                        }
+                        old_locale = uselocale (loc);
+                }
+
+                retval = dgettext ("gnome-desktop-3.0", g_hash_table_lookup (modifiers_map, modifier));
+
+                if (translation != NULL) {
+                        uselocale (old_locale);
+                        freelocale (loc);
+                }
+        } else {
+                retval = modifier;
+        }
+
+        g_hash_table_destroy (modifiers_map);
+
+        return g_strdup (retval);
+}
+
 /**
  * gnome_get_input_source_from_locale:
  * @locale: a locale string
diff --git a/libgnome-desktop/gnome-languages.h b/libgnome-desktop/gnome-languages.h
index 407a8c3..97f4af5 100644
--- a/libgnome-desktop/gnome-languages.h
+++ b/libgnome-desktop/gnome-languages.h
@@ -47,6 +47,8 @@ char *        gnome_get_language_from_code      (const char *code,
                                                  const char *translation);
 char *        gnome_get_country_from_code       (const char *code,
                                                  const char *translation);
+char *        gnome_get_translated_modifier     (const char *modifier,
+                                                 const char *translation);
 gboolean      gnome_get_input_source_from_locale (const char  *locale,
                                                   const char **type,
                                                   const char **id);
